zcash_primitives/src/transaction/fees/zip317.rs

```diff
@@ -0,0 +1,101 @@
+//! A [`FeeRule`] implmentation that implements [ZIP 317] fee calculation.
+//!
+//! [`FeeRule`]: crate::transaction::fees::FeeRule
+//! [ZIP 317]: https//zips.z.cash/zip-0317
+use core::cmp::max;
+
+use crate::{
+    consensus::{self, BlockHeight},
+    transaction::components::{
+        amount::{Amount, BalanceError},
+        transparent::fees as transparent,
+    },
+};
+
+/// A [`FeeRule`] implementation that implements the [ZIP 317] fee rule.
+///
+/// [`FeeRule`]: crate::transaction::fees::FeeRule
+/// [ZIP 317]: https//zips.z.cash/zip-0317
+#[derive(Clone, Debug)]
+pub struct FeeRule {
+    marginal_fee: Amount,
+    grace_actions: usize,
+    p2pkh_standard_input_size: usize,
+    p2pkh_standard_output_size: usize,
+}
+
+impl FeeRule {
+    /// Construct a new FeeRule instance with the specified parameter values.
+    /// Returns `None` if either `p2pkh_standard_input_size` or `p2pkh_standard_output_size`
+    /// are zero.
+    pub fn new(
```

Comment:
Is there a benefit to enabling non-standard parameters to be used? If so, `FeeRule::new` is too innocuous for such a constructor. We should rename and document it appropriately:
```suggestion
    pub fn non_standard(
```